skrobot/planner/sqp_based.py

```diff
@@ -0,0 +1,176 @@
+import numpy as np
+import scipy
+
+try:
+    # in python3
+    from functools import lru_cache
+except ImportError:
+    # in python2
+    from repoze.lru import lru_cache
+
+from skrobot.planner.utils import scipinize
+
+
+def sqp_plan_trajectory(collision_checker,
+                        av_start,
+                        av_goal,
+                        joint_list,
+                        n_wp,
+                        safety_margin=1e-2,
+                        with_base=False,
+                        weights=None,
+                        initial_trajectory=None,
+                        slsqp_option=None
+                        ):
+    """Gradient based trajectory optimization using scipy's SLSQP.
+
+    Collision constraint is considered in an inequality constraits.
+    Terminal constraint (start and end) is considered as an
+    equality constraint.
+
+    Parameters
+    ----------
+    av_start : numpy.array(n_dof,)
+        joint angle vector at start point
+    av_start : numpy.array(n_dof,)
+        joint angle vector at goal point
+    joint_list : list[skrobot.model.Link]
+        link list to be controlled (similar to inverse_kinematics function)
+    n_wp : int
+        number of waypoints
+    safety_margin : float
+        safety margin in collision checking
+    with_base: bool
+        If `with_base=False`, `n_dof` is the number of joints `n_joint`,
+        but if `with_base=True`, `n_dof = len(joint_list) + 3`.
+    weights : Union[numpy.array(n_dof,), None]
```

Comment:
What is the Union?

Reply:
following this : https://stackoverflow.com/questions/34647966/how-to-express-multiple-types-for-a-single-parameter-or-a-return-value-in-docstr

Reply:
We are using numpy style for docstrings. `numpy.ndarray or None` is suitable.

Reply:
fixed in https://github.com/iory/scikit-robot/pull/181/commits/66a1bd946c7bfeb1cda94b7bf780461877b62d16